proposals/bootstrapper-proposal.md

```diff
@@ -0,0 +1,253 @@
+# Kubeflow Bootstrapping Modifications
+
+## Motivation
+
+### Current Kubeflow bootstrapper requires cluster-admin privileges
+The current bootstrapper generates a kubeflow application into a PersistentVolume and will optionally deploy it to the API-server. The bootstrapper uses ksonnet to generate the kubernetes manifests that define the kubeflow runtime. The manifest generation by ksonnet is ordered across phases:
+- optional creation of a Namespace
```

Comment:
i think other than the namespace and PV creation, RBAC rules can be scoped to the namespace and deployment is possible w/ no cluster level bindings.

Reply:
Good point. Although I think there's work to be done to support multiple Kubeflow deployments per cluster.

CRD's are cluster scoped so we can only have 1 CRD per cluster. But CRD's are versioned so in theory all Kubeflow deployments using the same version of a CRD should be using the same CRD spec.

The controllers themselves can be scoped to only act on resources for a specific version and namespace so we could potentially support only claiming resources in a particular namespace. (This is an issue for our E2E tests as well).

There's also the question of ingress and Ambassador. Setting up ingress should probably require admin privileges.  

Would each deployment have its own use case?

It might be useful to try to define concrete user scenarios so we can figure out how to support them.

For example, the scenario we currently handle today is one cluster shared by a team and everyone uses a service account owned by the team to do work.

One use case we don't handle well today is when you have a team of datascientists each with access to different datasets. I think that might be the use case you are getting at because we would need a way to isolate each datascientist.

Reply:
@jlewi 
> Would each deployment have its own use case?

yes

> One use case we don't handle well today is when you have a team of datascientists each with access to different datasets. I think that might be the use case you are getting at because we would need a way to isolate each datascientist.

yes this is one of the use cases. We've had feedback from data scientists that there are common use cases where they may need to limit visibility to specific datasets due to [HIPPA](http://www.onlyhealthy.com/understanding-the-hipaa-law/) or other constraints.